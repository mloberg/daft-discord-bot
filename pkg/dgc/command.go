package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
