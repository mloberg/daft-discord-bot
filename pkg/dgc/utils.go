package dgc

import "github.com/bwmarrin/discordgo"

var InteractionResponseEphemeral uint64 = 1 << 6

// GetUser returns the user that created the interaction either in a DM or guild
func GetUser(i *discordgo.InteractionCreate) *discordgo.User {
	if i.Member != nil {
		return i.Member.User
	}

	return i.User
}
