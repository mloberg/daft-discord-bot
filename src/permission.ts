import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
