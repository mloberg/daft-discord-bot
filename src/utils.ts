export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
