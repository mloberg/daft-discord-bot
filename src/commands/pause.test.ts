import { Client, Guild, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import { FriendlyError } from '../error';
import { hasPermission } from '../permission';
import command from './pause';

const mocks = {
    react: jest.fn(),
    join: jest.fn(),
    pause: jest.fn(),
    permission: mocked(hasPermission),
};

jest.mock('discord.js', () => ({
    Client: jest.fn(),
    Guild: jest.fn(),
    TextChannel: jest.fn(),
    Message: jest.fn().mockImplementation(() => ({
        member: {
            voice: {
                channel: {
                    join: mocks.join.mockImplementation(() => ({
                        dispatcher: {
                            pause: mocks.pause,
                        },
                    })),
                },
            },
        },
        react: mocks.react,
    })),
}));

jest.mock('../permission');

describe('_pause configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('pause');
        expect(command.description).toEqual('Pause the playlist');
    });

    it('should have no aliases', () => {
        expect(command.alias).toBeUndefined();
    });
});

describe('_pause', () => {
    const client = new Client();
    const guild = new Guild(client, {});
    const channel = new TextChannel(guild, {});

    beforeEach(() => {
        mocks.react.mockClear();
        mocks.react.mockReturnThis();
        mocks.join.mockClear();
        mocks.pause.mockClear();
        mocks.permission.mockClear();
    });

    it('pauses current song', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: [], $0: 'pause' });

        expect(mocks.pause).toHaveBeenCalledTimes(1);
        expect(mocks.react).toBeCalledWith('ðŸŽ¶');
    });

    it('will not pause if not playing', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);
        mocks.join.mockReturnValue({ dispatcher: null });

        await command.run(message, { _: [], $0: 'pause' });

        expect(mocks.pause).toHaveBeenCalledTimes(0);
        expect(mocks.react).toHaveBeenCalledTimes(0);
    });

    it('will throw an error if not in a voice channel', async () => {
        (mocked(Message) as jest.Mock).mockImplementationOnce(() => {
            return { member: { voice: { channe: null } } };
        });
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        try {
            await command.run(message, { _: [], $0: 'pause' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('You are not in a voice channel');
        }
    });

    it('will throw an error if does not have role', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(false);

        try {
            await command.run(message, { _: [], $0: 'pause' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('You do not have permission to do that.');
        }
    });
});
