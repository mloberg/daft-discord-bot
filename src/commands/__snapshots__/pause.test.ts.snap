// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_pause is a command 1`] = `
Object {
  "description": "Pause the playlist",
  "name": "pause",
  "run": [Function],
}
`;
